I hit this issue again and i have now been able to test this PR. Before the boot entries are not created and the script succeeds:
```
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device: '/nix/store/kxw13f32nw9hsy4nmfhrc4vj97mr5619-initrd-linux-5.15.94/initrd' -> '/boot/efi/nixos/kxw13f32nw9hsy4nmfhrc4vj97mr5619-initrd-linux-5.15.94-initrd.efi'
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device: '/nix/store/4lqix0wdg2pbn1xqr1qi416bk4xi60ni-linux-5.15.104/bzImage' -> '/boot/efi/nixos/4lqix0wdg2pbn1xqr1qi416bk4xi60ni-linux-5.15.104-bzImage.efi'
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device: '/nix/store/mk7djkygw9xyv04xhxm0gy6p2h4gwd79-initrd-linux-5.15.104/initrd' -> '/boot/efi/nixos/mk7djkygw9xyv04xhxm0gy6p2h4gwd79-initrd-linux-5.15.104-initrd.efi'
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device
ignoring default profile in the list of boot entries because of the following error:
[Errno 28] No space left on device
activating the configuration...
setting up /etc...
```

With this patch, it fails with:
```
Traceback (most recent call last):
  File "/nix/store/igngsnrm6anqm4hcqr79k0x50zg22gij-systemd-boot", line 319, in <module>
    main()
  File "/nix/store/igngsnrm6anqm4hcqr79k0x50zg22gij-systemd-boot", line 288, in main
    raise e
  File "/nix/store/igngsnrm6anqm4hcqr79k0x50zg22gij-systemd-boot", line 277, in main
    write_entry(*gen, machine_id)
  File "/nix/store/igngsnrm6anqm4hcqr79k0x50zg22gij-systemd-boot", line 110, in write_entry
    initrd = copy_from_profile(profile, generation, specialisation, "initrd")
  File "/nix/store/igngsnrm6anqm4hcqr79k0x50zg22gij-systemd-boot", line 83, in copy_from_profile
    copy_if_not_exists(store_file_path, "/boot%s" % (efi_file_path))
  File "/nix/store/igngsnrm6anqm4hcqr79k0x50zg22gij-systemd-boot", line 28, in copy_if_not_exists
    shutil.copyfile(source, dest)
  File "/nix/store/hhk4wr7hwry854sq69chmrjqyi964p7y-python3-3.10.9/lib/python3.10/shutil.py", line 267, in copyfile
    _fastcopy_sendfile(fsrc, fdst)
  File "/nix/store/hhk4wr7hwry854sq69chmrjqyi964p7y-python3-3.10.9/lib/python3.10/shutil.py", line 156, in _fastcopy_sendfile
    raise err from None
  File "/nix/store/hhk4wr7hwry854sq69chmrjqyi964p7y-python3-3.10.9/lib/python3.10/shutil.py", line 142, in _fastcopy_sendfile
    sent = os.sendfile(outfd, infd, offset, blocksize)
OSError: [Errno 28] No space left on device: '/nix/store/kxw13f32nw9hsy4nmfhrc4vj97mr5619-initrd-linux-5.15.94/initrd' -> '/boot/efi/nixos/kxw13f32nw9hsy4nmfhrc4vj97mr5619-initrd-linux-5.15.94-initrd.efi'
```

> Could the test be even automated in nixos/tests/systemd-boot.nix ?

@raitobezarius I'm not able to find time to work on testing this part :( However, i think this patch improves the current situation where the `switch-to-configuration` script succeeds while the system won't be up to date on the next reboot...

--- a/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
+++ b/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
@@ -302,8 +302,12 @@ def main() -> None:
             if is_default:
                 write_loader_conf(*gen)
         except OSError as e:
-            profile = f"profile '{gen.profile}'" if gen.profile else "default profile"
-            print("ignoring {} in the list of boot entries because of the following error:\n{}".format(profile, e), file=sys.stderr)
+            # See https://github.com/NixOS/nixpkgs/issues/114552
+            if e.errno == errno.EINVAL:
+                profile = f"profile '{gen.profile}'" if gen.profile else "default profile"
+                print("ignoring {} in the list of boot entries because of the following error:\n{}".format(profile, e), file=sys.stderr)
+            else:
+                raise e
 
     for root, _, files in os.walk('@efiSysMountPoint@/efi/nixos/.extra-files', topdown=False):
         relative_root = root.removeprefix("@efiSysMountPoint@/efi/nixos/.extra-files").removeprefix("/")

